quickjs.c

```diff
@@ -57764,6 +57766,236 @@ static void _JS_AddIntrinsicCallSite(JSContext *ctx)
+                                            bool add_backtrace)
+{
+    JSDOMExceptionData *s;
+    JSValue obj, message, name;
+
+    obj = js_create_from_ctor(ctx, new_target, JS_CLASS_DOM_EXCEPTION);
+    if (JS_IsException(obj))
+        return JS_EXCEPTION;
+    if (!JS_IsUndefined(argv[0]))
+        message = JS_ToString(ctx, argv[0]);
+    else
+        message = JS_NewString(ctx, "");
+    if (JS_IsException(message))
+        goto fail1;
+    if (!JS_IsUndefined(argv[1]))
+        name = JS_ToString(ctx, argv[1]);
+    else
+        name = JS_NewString(ctx, "Error");
+    if (JS_IsException(name))
+        goto fail2;
+    s = js_malloc(ctx, sizeof(*s));
+    if (!s)
+        goto fail3;
+    s->name = name;
+    s->message = message;
+    s->code = -1;
+    JS_SetOpaqueInternal(obj, s);
+    if (add_backtrace)
+        build_backtrace(ctx, obj, JS_UNDEFINED, NULL, 0, 0,
+                        JS_BACKTRACE_FLAG_SKIP_FIRST_LEVEL);
+    return obj;
+fail3:
+    JS_FreeValue(ctx, name);
+fail2:
+    JS_FreeValue(ctx, message);
+fail1:
+    JS_FreeValue(ctx, obj);
+    return JS_EXCEPTION;
+}
+
+static JSValue js_domexception_constructor(JSContext *ctx, JSValueConst new_target,
+                                           int argc, JSValueConst *argv)
+{
+    if (JS_IsUndefined(new_target))
+        return JS_ThrowTypeError(ctx, "constructor requires 'new'");
+    return js_domexception_constructor0(ctx, new_target, argc, argv, true);
+}
+
+static JSValue js_domexception_get_name(JSContext *ctx, JSValueConst this_val)
+{
+    JSDOMExceptionData *s;
+
+    s = JS_GetOpaque2(ctx, this_val, JS_CLASS_DOM_EXCEPTION);
+    if (!s)
+        return JS_EXCEPTION;
+    return js_dup(s->name);
+}
+
+static JSValue js_domexception_get_message(JSContext *ctx, JSValueConst this_val)
+{
+    JSDOMExceptionData *s;
+
+    s = JS_GetOpaque2(ctx, this_val, JS_CLASS_DOM_EXCEPTION);
+    if (!s)
+        return JS_EXCEPTION;
+    return js_dup(s->message);
+}
+
+static JSValue js_domexception_get_code(JSContext *ctx, JSValueConst this_val)
+{
+    JSDOMExceptionData *s;
+    const char *name, *it;
+    int i;
+    size_t len;
+
+    s = JS_GetOpaque2(ctx, this_val, JS_CLASS_DOM_EXCEPTION);
+    if (!s)
+        return JS_EXCEPTION;
+    if (s->code == -1) {
+        name = JS_ToCStringLen(ctx, &len, s->name);
+        if (!name)
+            return JS_EXCEPTION;
+        for (i = 0; i < countof(js_dom_exception_names_table); i++) {
+            it = js_dom_exception_names_table[i].name;
+            if (it && !strcmp(it, name) && len == strlen(it)) {
+                s->code = i;
+                break;
+            }
+        }
+        s->code++;
+        JS_FreeCString(ctx, name);
+    }
+    return js_int32(s->code);
+}
+
+static const JSCFunctionListEntry js_domexception_proto_funcs[] = {
+    JS_CGETSET_DEF("name", js_domexception_get_name, NULL ),
+    JS_CGETSET_DEF("message", js_domexception_get_message, NULL ),
```

Comment:
These aren't wrong but if you want to DRY the code, you can do:
```suggestion
    JS_CGETSET_MAGIC_DEF("name", js_domexception_getfield, NULL,
                         offsetof(JSDOMExceptionData, name)),
    JS_CGETSET_MAGIC_DEF("message", js_domexception_getfield, NULL,
                         offsetof(JSDOMExceptionData, message)),
```
See js_callsite_getfield for an example:
```c
static JSValue js_callsite_getfield(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv, int magic)
{
    JSCallSiteData *csd = JS_GetOpaque2(ctx, this_val, JS_CLASS_CALL_SITE);
    if (!csd)
        return JS_EXCEPTION;
    JSValue *field = (void *)((char *)csd + magic);
    return js_dup(*field);
}
```
This is another one of those things we could generalize more because ultimately it just computes:
```c
JSValue *field = (void *)((char *)JS_VALUE_GET_OBJ(this_val)->u.opaque + magic);
```

Reply:
OK, done.  (At this scale though, it seems a bit closer to golfing than DRY :P)

Reply:
Not even code golfing, I'm afraid... I tried adding a generalized getter to the whole code base and it added about as many lines of code as it shaved off. Not worth it at the moment, I guess.